build.fsx

```diff
@@ -1,24 +1,129 @@
+            return! go ()
+        }
+        let cts = new CancellationTokenSource ()
+
+        let task = Async.StartAsTask (go (), cancellationToken = cts.Token)
+
+        interface IDisposable with
+            member _.Dispose () =
+                cts.Cancel ()
+                try
+                    task.Wait ()
+                with
+                | e when containsTaskCancellation e -> ()
+
+    let runProcessWithOutput (env : Map<string, string>) (command : string) (args : string seq) : string =
+        let psi = ProcessStartInfo (command, args)
+        psi.RedirectStandardOutput <- true
+        psi.RedirectStandardError <- true
+        env
+        |> Map.iter (fun key value ->
+            psi.EnvironmentVariables.[key] <- value
+        )
+        use proc = new Process ()
+        proc.StartInfo <- psi
+        // have to do the dance to avoid the classic deadlock condition where the subprocess is waiting to push
+        // to a full output buffer, but we aren't doing anything to clear the output buffer until the subprocess
+        // quits
+        let stdoutArr = ResizeArray ()
+        proc.OutputDataReceived.Add (fun evt -> lock stdoutArr (fun () -> stdoutArr.Add evt.Data))
+        let stderrArr = ResizeArray ()
+        proc.ErrorDataReceived.Add (fun evt -> lock stderrArr (fun () -> stderrArr.Add evt.Data))
+
+        if not (proc.Start ()) then
+            let args = args |> String.concat " "
+            failwith $"Failed to start process '%s{command}' with args: %s{args}"
+        proc.BeginOutputReadLine ()
+        proc.BeginErrorReadLine ()
+        Console.Write $"(started process %i{proc.Id}) "
+
+        do
+            use waiter = new Waiter (stdoutArr, stderrArr)
+            proc.WaitForExit ()
+
+        let stdout = stdoutArr |> Seq.filter (fun s -> not (String.IsNullOrEmpty s)) |> String.concat " "
+
+        if proc.ExitCode <> 0 then
+            let args = args |> String.concat " "
+            let stdout = if String.IsNullOrEmpty stdout then "" else $"\nStdout:\n  %s{stdout}"
+            failwith $"Process '%s{command}' failed with nonzero exit code %i{proc.ExitCode}. Args: %s{args}.%s{stdout}"
+
+        stdout
+
+    let inline runProcessWithEnv env command args : unit = runProcessWithOutput env command args |> ignore<string>
+    let inline runProcess command args : unit = runProcessWithEnv Map.empty command args
+
+    let rec copyDir (source : DirectoryInfo) (target : DirectoryInfo) : unit =
+        target.Create ()
+        for file in source.EnumerateFiles () do
+            file.CopyTo (Path.Combine (target.FullName, file.Name)) |> ignore<FileInfo>
+        for dir in source.EnumerateDirectories () do
+            copyDir dir (Path.Combine (target.FullName, dir.Name) |> DirectoryInfo)
+
+// --------------------------------------------------------------------------------------
+// Clean build results
+
+type HaveCleaned = HaveCleaned
```

Comment:
Pattern: each FAKE stage is now represented by a token indicating that you have run that stage. A dependency is now encoded as the function taking that token as an argument.